language/stdlib/modules/libra_system.move

```diff
@@ -310,46 +280,32 @@ module LibraSystem {
         !Vector::is_empty(&get_validator_index_(validators_vec_ref, addr))
     }
 
-    // Discovery methods
-
     fun emit_discovery_set_change() acquires DiscoverySet {
         let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
         Event::emit_event<DiscoverySetChangeEvent>(
             &mut discovery_set_ref.change_events,
             DiscoverySetChangeEvent {
                 new_discovery_set: *&discovery_set_ref.discovery_set,
+                new_validator_set: get_validator_set(),
             },
         );
     }
 
-    public fun get_discovery_address(d: &DiscoveryInfo): address {
-        d.addr
-    }
+    // Functions for testing
 
-    public fun get_discovery_config(d: &DiscoveryInfo): &ValidatorConfig::DiscoveryConfig {
-        &d.discovery_config
-    }
-
-    // Get the DiscoveryInfo for the ith validator
-    public fun get_ith_discovery_info(i: u64): DiscoveryInfo acquires DiscoverySet {
-        let discovery_vec_ref = &borrow_global<DiscoverySet>(0xD15C0).discovery_set;
-        *Vector::borrow(discovery_vec_ref, i)
+    // This getter is only used in tests
+    public fun get_validator_config(addr: address): ValidatorConfig::Config {
+        let validator_set = get_validator_set();
+        let validator_index_vec = get_validator_index_(&validator_set.validators, addr);
+        *&(*Vector::borrow(&validator_set.validators, *Vector::borrow(&validator_index_vec, 0))).config
     }
 
-    // Get the index of the discovery info with address `addr` in `discovery_set`.
-    // Aborts if `addr` is not in discovery set.
-    public fun get_discovery_index(discovery_set: &vector<DiscoveryInfo>, addr: address): u64 {
-        let len = Vector::length(discovery_set);
-        let i = 0;
-        loop {
-            if (get_discovery_address(Vector::borrow(discovery_set, i)) == addr) {
-                return i
-            };
-
-            i = i + 1;
-            if (i >= len) { break };
-        };
-
-        abort 99
+    // This getter is only used in tests
+    public fun get_fullnode_config(addr: address): ValidatorConfig::FullNodeConfig acquires DiscoverySet {
+        // To simplify the code, we use the fact that at the moment fullnode config is stored at
+        // the same index in the vector of fullnode configs as in the vector of validator configs
+        let discovery_set_ref = &borrow_global<DiscoverySet>(0xD15C0).discovery_set;
+        let validator_index_vec = get_validator_index_(&get_validator_set().validators, addr);
```

Comment:
this seems dangerous since there's no guarantee that discovery set is in the same order as the validator set

Reply:
The way the add_validator and remove_validator code is currently written, it is guaranteed that the validator config and the fullnode config will have the same index in their corresponding sets, they are in full correspondence. But I see your point in that this might change! Removing this assumption will add about ~30 lines of code to the module (basically copy pasting the function `get_validator_index_` into the `get_full_node_index_` function). But I thought, perhaps it's best to add those lines later on when we break the 1-to-1 correspondence between the fullnodes and the validators? For now I can assert here that the fullnode address at this position is an expected address.